src/ring_theory/adjoin/power_basis.lean

```diff
@@ -71,3 +72,122 @@ open algebra
+begin
+  intro i,
+  let Q := (X ^ n) %ₘ (minpoly R B.gen),
+  have : B.gen ^ n = aeval B.gen Q,
+  { rw [← @aeval_X_pow R _ _ _ _ B.gen, ← mod_by_monic_add_div (X ^ n) (minpoly.monic hB)],
+    simp },
+  by_cases hQ : Q = 0,
+  { simp [this, hQ, is_integral_zero] },
+  have hlt : Q.nat_degree < B.dim,
+  { rw [← B.nat_degree_minpoly, hmin, (minpoly.monic hB).nat_degree_map,
+      nat_degree_lt_nat_degree_iff hQ],
+    letI : nontrivial R := nontrivial.of_polynomial_ne hQ,
+    exact degree_mod_by_monic_lt _ (minpoly.monic hB),
+    apply_instance },
+  rw [this, aeval_eq_sum_range' hlt],
+  simp only [linear_equiv.map_sum, linear_equiv.map_smulₛₗ, ring_hom.id_apply, finset.sum_apply'],
+  refine is_integral.sum _ (λ j hj, _),
+  replace hj := finset.mem_range.1 hj,
+  rw [← fin.coe_mk hj, ← B.basis_eq_pow, algebra.smul_def,
+    is_scalar_tower.algebra_map_apply R S A, ← algebra.smul_def, linear_equiv.map_smul],
+  simp only [algebra_map_smul, finsupp.coe_smul, pi.smul_apply, B.basis.repr_self_apply],
+  by_cases hij : (⟨j, hj⟩ : fin _) = i,
+  { simp only [hij, eq_self_iff_true, if_true],
+    rw [algebra.smul_def, mul_one],
+    exact is_integral_algebra_map },
+  { simp [hij, is_integral_zero] }
+end
+
+variable {B}
+
+/-- Let `B : power_basis S A` be such that `is_integral R B.gen`, and let `x y : A` be elements with
+integral coordinates in the base `B.basis`. Then `is_integral R ((B.basis.repr (x * y) i)` for all
+`i` if `minpoly S B.gen = (minpoly R B.gen).map (algebra_map R S)`. This is the case if `R` is a GCD
+domain and `S` is its fraction ring. -/
+lemma repr_mul_is_integral [is_domain S] {x y : A} (hx : ∀ i, is_integral R (B.basis.repr x i))
+  (hy : ∀ i, is_integral R (B.basis.repr y i))
+  (hmin : minpoly S B.gen = (minpoly R B.gen).map (algebra_map R S)) :
+  ∀ i, is_integral R ((B.basis.repr (x * y) i)) :=
+begin
+  intro i,
+  rw [← B.basis.sum_repr x, ← B.basis.sum_repr y, finset.sum_mul_sum, linear_equiv.map_sum,
+    finset.sum_apply'],
+  refine is_integral.sum _ (λ I hI, _),
+  simp only [algebra.smul_mul_assoc, algebra.mul_smul_comm, linear_equiv.map_smulₛₗ,
+    ring_hom.id_apply, finsupp.coe_smul, pi.smul_apply, id.smul_eq_mul],
+  refine is_integral_mul (hy _) (is_integral_mul (hx _) _),
+  simp only [coe_basis, ← pow_add],
+  refine repr_gen_pow_is_integral hB hmin _ _,
+end
+
+/-- Let `B : power_basis S A` be such that `is_integral R B.gen`, and let `x : A` be and element
+with integral coordinates in the base `B.basis`. Then `is_integral R ((B.basis.repr (x ^ n) i)` for
+all `i` and all `n` if `minpoly S B.gen = (minpoly R B.gen).map (algebra_map R S)`. This is the case
+if `R` is a GCD domain and `S` is its fraction ring. -/
+lemma repr_pow_is_integral [is_domain S] {x : A} (hx : ∀ i, is_integral R (B.basis.repr x i))
+  (hmin : minpoly S B.gen = (minpoly R B.gen).map (algebra_map R S)) (n : ℕ) :
+  ∀ i, is_integral R ((B.basis.repr (x ^ n) i)) :=
+begin
+  by_cases htriv : nontrivial A, swap,
```

Comment:
Do you have any idea why the nontriviality tactic doesn't do this bit for you? Would be good to diagnose that if possible and make the tactic better (maybe just by adding a lemma or two).

Reply:
I don't know how the tactic works, but it has to do `intro i`, realize that `x ^ n = 0`, apply a bunch of `simp` lemma and use `is_integral_zero`. Even trying `nontriviality A using [is_integral_zero]` after `intro i` fails, so I think it does not realize that `x ^ n = 0`.